src/middlewared/middlewared/utils/threading.py

```diff
@@ -31,21 +45,30 @@ def __init__(self):
         self.executor = ThreadPoolExecutor(
             self.thread_count,
             "IoThread",
-            initializer=lambda: set_thread_name("IoThread"),
+            initializer=initializer,
+            initargs=("IoThread", thread_local_storage),
         )
 
     def submit(self, fn, *args, **kwargs):
-        if len(self.executor._threads) == self.thread_count and self.executor._idle_semaphore._value - 1 <= 1:
+        if (
+            len(self.executor._threads) == self.thread_count
+            and self.executor._idle_semaphore._value - 1 <= 1
+        ):
             fut = Future()
             logger.trace("Calling %r in a single-use thread", fn)
-            start_daemon_thread(name=f"ExtraIoThread_{next(counter)}", target=worker, args=(fut, fn, args, kwargs))
+            start_daemon_thread(
+                name=f"ExtraIoThread_{next(counter)}",
+                target=worker,
+                args=(fut, fn, thread_local_storage, args, kwargs),
+            )
             return fut
 
         return self.executor.submit(fn, *args, **kwargs)
 
 
-def worker(fut, fn, args, kwargs):
+def worker(fut, fn, tls, args, kwargs):
     set_thread_name("ExtraIoThread")
```

Comment:
Can we just call `initializer` here to avoid code duplication?

Reply:
Yep, I can do that.